### PR TITLE
Update event push action and receipt tables to support threads.

### PR DESCRIPTION
**Pull request by clokep**
This is kind of split out of #13181 / #13202; it includes the database schema deltas only, with minor changes to the queries so we start inserting the correct values, and a background job to backfill data.

The overall design of this is to add a `thread_id` column to the `event_push_actions` (+ `event_push_actions_staging`) and `event_push_summary` tables. This allows the results to be segmented by the thread (by storing either the root event ID in the `thread_id` column or the special value `"main"`).

`receipts_linearized` and `receipts_graph` also gain a `thread_id` column, which mostly works the same, but `NULL` is a special value meaning the receipt is "unthreaded".

See https://github.com/matrix-org/synapse/pull/13181#discussion_r963355433

After this we'll need to do a follow-up to actually use the new data we're inserting (and eventually remove the old indices).

For deployment / rollout purposes:

1. This adds the column(s) (and starts putting data into it).
2. PR2 will require `thread_id` on the `event_push_*` tables. (Which will require a schema bump?) At this point, we can do selects and joins properly for notification counts in threads. We'll drop the old indexes on `(user_id, room_id)`.
3. PR3 will drop the constraints for `receipts_*` tables on `(user_id, room_id, receipt_type)` and update queries to actually use the `thread_id` field.

Part of #12550; implements the storage for MSC3771 and MSC3773.

### CODE REVIEW

File: synapse/storage/databases/main/events.py [hunk 1]
```diff
@@ -2192,9 +2192,9 @@ def _set_push_actions_for_event_and_users_txn(
         sql = """
             INSERT INTO event_push_actions (
                 room_id, event_id, user_id, actions, stream_ordering,
-                topological_ordering, notif, highlight, unread
+                topological_ordering, notif, highlight, unread, thread_id
             )
-            SELECT ?, event_id, user_id, actions, ?, ?, notif, highlight, unread
+            SELECT ?, event_id, user_id, actions, ?, ?, notif, highlight, unread, thread_id
```
Comment (clokep) [inline synapse/storage/databases/main/events.py:2197]: Maybe we should be providing the `thread_id` like we do the `room_id`? (I.e. not storing it in `event_push_actions_staging` and only storing it in `event_push_actions`)

File: synapse/storage/schema/main/delta/72/07thread_receipts.sql.postgres [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- Add a nullable column for thread ID to the receipts table; this allows a
+-- receipt per user, per room, as well as an unthreaded receipt (corresponding
+-- to a null thread ID).
```
Comment (clokep) [inline synapse/storage/schema/main/delta/72/07thread_receipts.sql.postgres:18]: Do we need to do anything with `SCHEMA_VERSION` here? (Or note that we added this or anything?)

**Review by erikjohnston — APPROVED** [review]

File: synapse/storage/schema/main/delta/72/07thread_receipts.sql.postgres (hunk 2 above)
Comment (clokep) [reply synapse/storage/schema/main/delta/72/07thread_receipts.sql.postgres]: From discussion it seems safest that we actually bump to 73.

**Comment by clokep** [comment]
I think this needs to take into account the changes from #13703.

**Pull request merged**